walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,100 @@
 
 package com.google.sps;
 
+import com.google.sps.Event;
+import com.google.sps.TimeRange;
+import com.google.sps.MeetingRequest;
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
 
 public final class FindMeetingQuery {
-  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+  public Collection<TimeRange> query(Collection<Event> events, 
+                                      MeetingRequest request) {
+    // Get unavailable time for all required attendees in meeting request
+    ArrayList<TimeRange> unavailable = new ArrayList<>();
+    for (String attendee: request.getAttendees()) {
+      ArrayList<TimeRange> meetingTimes = getMeetingTimes(events, attendee);
+      unavailable = merge(unavailable, meetingTimes);
+    }
+
+    Collection<TimeRange> result = new ArrayList<>();
+    for (TimeRange t: getAvailability(unavailable)) {
+      // Get time ranges that are longer than the requested meeting duration
+      if (t.duration() >= request.getDuration()) {
+        result.add(t);
+      }
+    }
+
+    return result;
+  }
+
+  private ArrayList<TimeRange> merge(ArrayList<TimeRange> result, 
+                                      ArrayList<TimeRange> other) {
+    if (result.isEmpty()) {
+      return other;
+    }
+    if (other.isEmpty()) {
+      return result;
+    }
+
+    result.addAll(other);
+    Collections.sort(result, TimeRange.ORDER_BY_START);
+
+    // Merge overlapping ranges
+    for (int i = 0; i < result.size()-1; i++) {
+      TimeRange current = result.get(i);
+      TimeRange next = result.get(i+1);
+
+      if (current.overlaps(next)) {
+        // Case 1:  |_______|
+        //            |__|
+        if (current.contains(next)) {
+          result.remove(next);
+          i--;
+        }
+        // Case 2:  |______|
+        //               |______|
+        else {
+          result.set(i, TimeRange.fromStartEnd(current.start(), 
+                                                next.end(), false));
+          result.remove(next);
+          i--;
+        }
+      }
+    }
+    
+    return result;
+  }
+
+  /**
+   * Gets an attendee's unavailable time ranges
+   * Assume required meetings for one person do not overlap
+   */
+  private ArrayList<TimeRange> getMeetingTimes(Collection<Event> events, 
+                                                String attendee) {
+    ArrayList<TimeRange> meetingTimes = new ArrayList<>();
+    for (Event e: events) {
+      if (e.getAttendees().contains(attendee)) {
+        meetingTimes.add(e.getWhen());
+      }
+    }
+
+    return meetingTimes;
+  }
+
+  /**
+   * Returns attendee's available time ranges given unavailable time ranges
+   */
+  private Collection<TimeRange> getAvailability(Collection<TimeRange> unavailable) {
```

Comment:
This implementation relies on the unavailable is sorted.
please document this and make it clear either via the function name or the parameter name.